src/axom/quest/SamplingShaper.hpp

```diff
@@ -108,10 +116,29 @@ class InOutSampler
     m_octree->generateIndex();
   }
 
-  void sampleInOutField(mfem::DataCollection* dc,
-                        shaping::QFunctionCollection& inoutQFuncs,
-                        int sampleRes)
+  template <int FromDim, int ToDim>
+  std::enable_if_t<ToDim != DIM, void> sampleInOutField(
+    mfem::DataCollection*,
+    shaping::QFunctionCollection&,
+    int,
+    PointProjector<FromDim, ToDim>)
+  {
+    static_assert(
+      ToDim != DIM,
+      "Do not call this function -- it only exists to appease the compiler!"
+      "Projector's return dimension (ToDim), must match class dimension (DIM)");
+  }
+
+  template <int FromDim, int ToDim = DIM>
+  std::enable_if_t<ToDim == DIM, void> sampleInOutField(
+    mfem::DataCollection* dc,
+    shaping::QFunctionCollection& inoutQFuncs,
+    int sampleRes,
+    PointProjector<FromDim, ToDim> projector = {})
```

Comment:
@publixsubfan (or others) -- is this the right form for supplying the projection callback? 
The alias is to a `std::function` (defined at the top of this function) and there is no captured data.

An alternate form might be to use the c-style `ToPoint (*)(FromPoint) = nullptr`, which might improve performance slightly at the expense of clarity (according to some online links). When I profiled the above code, there doesn't appear to be significant overhead for this form.

Reply:
I think this is fine, although a comment describing the constraints on the defaulted argument would be helpful if it's not  there already.

Reply:
Thanks -- I added docs for the two `sampleInOutField` overloads.

Reply:
If performance is the main concern, we should probably use a templated function argument, along the lines of `Func&& projector`; I think both `std::function` and a C-style function pointer would prevent inlining the underlying projector function, since they both introduce a layer of runtime indirection.

The only drawback would be that we don't have a good way of constraining the passed-in function argument unless we were to bump up to C++17 and do something along the lines of `static_assert(std::is_invocable<...>)`.

Reply:
Thanks @publixsubfan -- I'm not sure if it'll work in this case since `sampleInOutField` does not directly take a lambda, but rather a callback that users register before shaping. The latter is stored as a `std::function`.

I ran this through Allinea map in release mode and the function call did not show up, so I'm not particularly concerned about the performance for now. 
(In fact, the entire `runShapeQuery()` did not really show up in 2D for meshes w/ $16^3$ or $32^3$ query points per material. The dominant call by far was `adjustVolumeFractions` which performs FCT + a large matrix assembly and solve.)

Reply:
Gotcha -- in that case my personal preference would be to stick with the `std::function`, but either is probably fine.